### PR TITLE
Converted batch norm over to use symints

### PR DESCRIPTION
**Pull request by Chillee**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #84829
* #84830
* __->__ #84113

### CODE REVIEW

**Comment by Chillee** [comment]
Doesn't actually work yet due to operations between tensors and sizes.

File: test/test_dynamic_shapes.py [hunk 1]
```diff
@@ -134,7 +134,7 @@ def create_symbolic_tensor(name, arg, shape_env):
     return FakeSymbolicTensor(sym_shapes, sym_strides, arg.dtype, arg.layout, arg.requires_grad, arg.device)
 
 
-CPP_SYMINT_CLASS = type(torch._C.SymIntNode.new_symint(1))
+CPP_SYMINT_CLASS = type(torch.SymIntNode.new_symint(1))
```
Comment (wconstab) [inline test/test_dynamic_shapes.py:137]: how come we take the type of an instance instead of directly referring to a class?  I guess its either (a) its hard to refer to a class or (b) we expect different classes depending on which mode this test is run under?

File: torch/_subclasses/fake_tensor.py [hunk 2]
```diff
@@ -581,10 +581,12 @@ def __torch_dispatch__(self, func, types, args=(), kwargs=None):
                 if symbolic_shapes.is_symbolic_op(func):
                     return symbolic_shapes.handle_symbolic_op(func, args, kwargs)
                 if func == aten.size.default:
-                    raise RuntimeError(
+                    print(
                         "Trying to call aten.size on a tensor with symbolic shapes. "
                         "It's likely that this is from calling tensor.shape in C++"
                     )
+                    # We do this to allow for better error localization with `TORCH_SHOW_CPP_STACKTRACES=1`
```
Comment (wconstab) [inline torch/_subclasses/fake_tensor.py:588]: why does not raising here make it easier to find a stacktrace? does `return None` trigger someone else to raise, and their stacktrace is more useful?

File: torch/_meta_registrations.py [hunk 3]
```diff
@@ -344,7 +348,7 @@ def pick_memory_format():
 
     else:
         out_channels = weight.shape[0]
-        if weight.shape[1] != input_tensor.shape[1] / groups:
+        if weight.shape[1] * groups != input_tensor.shape[1]:
```
Comment (wconstab) [inline torch/_meta_registrations.py:351]: is this what fixed the xfails for conv above? it seems like the exception above is 'couldn't find a symbolic meta func'

File: test/test_proxy_tensor.py [hunk 4]
```diff
@@ -889,16 +889,11 @@ def f(a):
...
-    xfail('bool', ''),  # aten._to_copy.default - couldn't find symbolic meta function/decomposition
     xfail('broadcast_tensors', ''),  # aten.size.default - couldn't find symbolic meta function/decomposition
     xfail('bucketize', ''),  # aten.bucketize.Tensor - couldn't find symbolic meta function/decomposition
-    xfail('byte', ''),  # aten._to_copy.default - couldn't find symbolic meta function/decomposition
     xfail('cartesian_prod', ''),  # Tensors of type TensorImpl do not have numel
     xfail('cdist', ''),  # aten.size.default - couldn't find symbolic meta function/decomposition
-    xfail('chalf', ''),  # aten._to_copy.default - couldn't find symbolic meta function/decomposition
-    xfail('char', ''),  # aten._to_copy.default - couldn't find symbolic meta function/decomposition
     xfail('cholesky_solve', ''),  # Could not run 'aten::_cholesky_solve_helper' with arguments from the 'Meta' back...
     xfail('chunk', ''),  # aten.size.default - couldn't find symbolic meta function/decomposition
```
Comment (wconstab) [inline test/test_proxy_tensor.py:898]: so if i wanted to fix something, such as all the tests that complain about aten.size.default, how would i go about it?

File: torch/_meta_registrations.py (hunk 3 above)
Comment (Chillee) [reply torch/_meta_registrations.py:351]: Yeah I think so. And also the `copy_` meta registration.

Comment (Chillee) [reply torch/_meta_registrations.py:351]: The error messages might be out of date :P I can regenerate them.

File: test/test_dynamic_shapes.py (hunk 1 above)
Comment (Chillee) [reply test/test_dynamic_shapes.py:137]: I don't know, I just replaced all the code haha.

cc: @Krovatkin

File: test/test_proxy_tensor.py (hunk 4 above)
Comment (Chillee) [reply test/test_proxy_tensor.py:898]: So complaining about `aten.size.default` is a bit misleading, since complaining about `aten.size.default` just means that we're calling for concrete integers inside C++ somewhere.

File: torch/_meta_registrations.py [hunk 5]
```diff
@@ -87,6 +87,8 @@ def meta_fft_c2r(self, dim, normalization, lastdim):
 
 @register_meta(aten.copy_.default)
 def meta_copy_(self, src, non_blocking=False):
+    if src.device.type == "meta":
+        raise NotImplementedError("Cannot copy out")
```
Comment (wconstab) [inline torch/_meta_registrations.py:91]: maybe go the extra mile and say "Cannot copy data from a meta tensor"?

**Review by wconstab — APPROVED** [review]
LGTM

File: torch/_meta_registrations.py (hunk 5 above)
Comment (Chillee) [reply torch/_meta_registrations.py:91]: This is actually the current error message lol, and the test errors if I provide anything else.

File: aten/src/ATen/native/Normalization.cpp [hunk 6]
```diff
@@ -443,14 +444,14 @@ std::tuple<Tensor, Tensor, Tensor, Tensor, int64_t> _batch_norm_impl_index(
...
+  auto num_features = input.sym_sizes()[1];
 
-  if (input.numel() == 0) {
+  if (input.sym_numel() == 0) {
     Tensor reserve = at::empty({0}, input.options().dtype(kByte));
     auto options = input.options().dtype(
         at::toAccumulateType(input.scalar_type(), /*is_cuda=*/input.is_cuda()));
-    auto save_mean = at::empty({num_features}, options);
-    auto save_invstd = at::empty({num_features}, options);
+    auto save_mean = at::empty_symint(c10::SymIntArrayRef({num_features}), options);
```
Comment (ezyang) [inline aten/src/ATen/native/Normalization.cpp:453]: The ArrayRef here shouldn't be necessary

File: torch/_meta_registrations.py [hunk 7]
```diff
@@ -214,7 +221,6 @@ def _to_copy(
     device = self.device if device is None else device
     layout = self.layout if layout is None else layout
     assert pin_memory is None
-    assert memory_format is None
     return self.new_empty(self.shape, dtype=dtype, device=device, pin_memory=pin_memory)
```
Comment (ezyang) [inline torch/_meta_registrations.py:224]: No, you need this assert. Assert it's None or the default contiguity

File: torch/_subclasses/fake_tensor.py (hunk 2 above)
Comment (ezyang) [reply torch/_subclasses/fake_tensor.py:588]: A better fix for this is to localize Python errors that go back into C++. But at minimum, the print here should go to sys.stderr

**Review by ezyang — APPROVED** [review]

File: aten/src/ATen/native/Normalization.cpp (hunk 6 above)
Comment (Chillee) [reply aten/src/ATen/native/Normalization.cpp:453]: It doesn't seem like it supports implicit conversions.

File: torch/_meta_registrations.py (hunk 7 above)
Comment (Chillee) [reply torch/_meta_registrations.py:222]: What should I do if it's channels last? Manually create a `new_empty_strided`?

File: aten/src/ATen/native/Normalization.cpp (hunk 6 above)
Comment (ezyang) [reply aten/src/ATen/native/Normalization.cpp:453]: ^- @Krovatkin

File: torch/_meta_registrations.py (hunk 7 above)
Comment (ezyang) [reply torch/_meta_registrations.py:222]: If it's channels last then yes, we need to properly respect the requested memory format. It might be easiest to add memory format as an argument to `new_empty`, not sure. Are we using `new_empty` here for composite compliance? `torch.empty` would also let you directly use memory format.

Comment (Chillee) [reply torch/_meta_registrations.py:222]: Yeah lemme just fix it here: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/EmptyTensor.cpp#L329
